osu.Game/Screens/Play/OffsetCorrectionClock.cs

```diff
@@ -36,10 +37,31 @@ public override void ProcessFrame()
             updateOffset();
         }
 
-        private void updateOffset()
+        protected virtual void updateOffset()
         {
             // we always want to apply the same real-time offset, so it should be adjusted by the difference in playback rate (from realtime) to achieve this.
             base.Offset = Offset * Rate;
         }
+
+        protected void updateLatencyAssumption()
+        {
+            // The latency assumption baked into beatmaps should also be adjusted by the difference in playback rate.
+            // There should only be one clock that does this.
+            base.Offset += Editor.WAVEFORM_VISUAL_OFFSET * (Rate - 1.0);
```

Comment:
That just happens to be where the latency assumption offset is stored:

https://github.com/ppy/osu/blob/7bc571dd517bdcc5493eac8e28b2ea5309726591/osu.Game/Screens/Edit/Editor.cs#L63-L74

Reply:
well if it's going to be used for _actual_ audio _it shouldn't live there anymore and should no longer be called "visual"_

Reply:
Understood

Reply:
when the rate is slowed down, the value will be negative to account for the drag, and when sped up it will be positive to account for the rush.

Reply:
I'm thinking that the editor might still be the most appropriate place for that constant to be. The latency assumption offset exists based on how users create a beatmap, and how they create a beatmap is influenced by what they see in the editor.

The offset exists because the editor essentially tricks the user into thinking that a signal starts at a position 20ms earlier than where it actually starts.

I could change the name to something more general like `LATENCY_ASSUMPTION_OFFSET`

Reply:
> EDIT: sorry, you mean negative clock rate as in reversed playback?

correct.

Reply:
Alright, well how about if I move it to the `Beatmap` class? That seems appropriate as the context in how it's being used might be visual or audio-related but the one thing that's always true is its strong ties to beatmaps and how they're written.

Reply:
I could post a video showing the before and after. Would that work for you?

Reply:
if it clearly shows what the issue is and that this fixes it, sure

Reply:
The before starts at the beginning of the video, and the after starts at 0:45.
In the before, there is a noticeable delay between hit object sample effects and the track, but only at slower speeds. In the after, this delay is virtually gone.

I have a global audio offset of 5ms, I find that to be ever so slightly more accurate on my setup. I make sure to show at the beginning of both recordings that this global offset is consistent.

https://github.com/ppy/osu/assets/3401272/27b00a82-e9b3-44f5-acdb-6820665f1ce7

Reply:
I'm genuinely curious, I want to see how close my calculations are to what ends up actually happening in this example. Let's take a closer look at that hit object at 0:36:

![editor](https://github.com/ppy/osu/assets/3401272/bf9c98a9-01d2-41e9-82f9-2ad14b9258d7)

According to the editor, it starts at 14,440ms, which suggests that if we were to look at the track in Audacity, it would show that that beat/signal actually starts at 14,460ms. Let's double check:

![time](https://github.com/ppy/osu/assets/3401272/fd96cb81-77c6-4f9d-86fe-d5c41f7852ce)

Alright, now the playback rate at that time was 14%. So based on my calculations, the real-time delay between hit object and track downbeat should be `20 - 20 / 0.14 = −122.857142857`. Let's see if that holds true, looking now at the audio from the video at roughly 0:36:

![gap](https://github.com/ppy/osu/assets/3401272/fdbd6465-1490-42c9-ab6b-72b075bd4ba6)

Bear in mind that Audacity rounds to the nearest millisecond. Anyways, there you have it, the math, as far as I can tell, completely checks out in this particular case.